### PR TITLE
Add error handler for DenyAccessAuthenticator

### LINKED ISSUE #13368: Issue when using DenyAuthenticator in direct-grant flow
### Describe the bug

Hey there.

I'm currently trying to deny access to users that are not associated with a specific client role. Following the documentation on https://www.keycloak.org/docs/latest/server_admin/index.html#explicitly-deny-allow-access-in-conditional-flows the flow that I'm trying to use is the following (basically a copy of Direct Grant changing just the OTP conditional):
```
Username Validation   [REQUIRED]
Password   [REQUIRED]
Check Role   [CONDITIONAL]
  Condition - User Role(User MUST HAVE ROLE)   [REQUIRED]
  Deny Access   [REQUIRED]
```
When I try to authenticate a user that **has** the role, everything works as expected, however when I try to authenticate a user that doesn't have the role associated I get an **unknown_error** and the logs are as follows:
```
2022-07-27 12:03:59,948 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-54) Uncaught server error: java.lang.IllegalArgumentException: RESTEASY003715: path was null
        at org.jboss.resteasy.specimpl.ResteasyUriBuilderImpl.path(ResteasyUriBuilderImpl.java:382)
        at org.keycloak.authentication.AuthenticationProcessor$Result.getActionUrl(AuthenticationProcessor.java:562)
        at org.keycloak.authentication.AuthenticationProcessor$Result.form(AuthenticationProcessor.java:536)
        at org.keycloak.authentication.authenticators.access.DenyAccessAuthenticator.authenticate(DenyAccessAuthenticator.java:49)
        at org.keycloak.authentication.DefaultAuthenticationFlow.processSingleFlowExecutionModel(DefaultAuthenticationFlow.java:460)
        at org.keycloak.authentication.DefaultAuthenticationFlow.processFlow(DefaultAuthenticationFlow.java:264)
        at org.keycloak.authentication.DefaultAuthenticationFlow.processSingleFlowExecutionModel(DefaultAuthenticationFlow.java:395)
        at org.keycloak.authentication.DefaultAuthenticationFlow.processFlow(DefaultAuthenticationFlow.java:264)
        at org.keycloak.authentication.AuthenticationProcessor.authenticateOnly(AuthenticationProcessor.java:1030)
        at org.keycloak.protocol.oidc.endpoints.TokenEndpoint.resourceOwnerPasswordCredentialsGrant(TokenEndpoint.java:603)
        at org.keycloak.protocol.oidc.endpoints.TokenEndpoint.processGrantRequest(TokenEndpoint.java:200)
        at jdk.internal.reflect.GeneratedMethodAccessor343.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
        at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
        at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
        at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
        at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:192)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:152)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:183)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:141)
        at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:32)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
        at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
        at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
        at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
        at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
        at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:151)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:82)
        at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:42)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
        at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
        at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:67)
        at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:55)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
        at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder$5.handle(VertxHttpRecorder.java:380)
        at io.quarkus.vertx.http.runtime.VertxHttpRecorder$5.handle(VertxHttpRecorder.java:358)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
        at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
        at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$createBlockingHandler$1(QuarkusRequestFilter.java:71)
        at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:159)
        at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
        at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:157)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:829)

2022-07-27 12:03:59,951 WARN  [io.agroal.pool] (executor-thread-54) Datasource '<default>': JDBC resources leaked: 2 ResultSet(s) and 2 Statement(s)
```
Is this a bug or am I missing something?

### Version

18.0.2

### Expected behavior

Getting an unauthorized error when trying to authenticate a user that doesn't have the required role.

### Actual behavior

Getting a server error, with an exception on the logs.

### How to Reproduce?

- Create a copy of direct grant authentication flow, replace the OTP step with a conditional that checks if the user doesn't have the role and then denies access.
- Add this flow to a client
- create a user without the specified role and authenticate it.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by lexcao**
Closes #13368

These changes introduce a new error handler for building error based on the media type.
- It should create error form response when it is valid HTML request
- Otherwise, it could create error response with JSON

[open question]
* there is not support message bundle for JsonErrorHandler, should I add it as well?

@mposolda

### CODE REVIEW

File: services/src/main/java/org/keycloak/authentication/authenticators/access/DenyAccessAuthenticator.java [hunk 1]
```diff
@@ -46,9 +46,12 @@ public void authenticate(AuthenticationFlowContext context) {
                 .orElse(Messages.ACCESS_DENIED);
 
         context.getEvent().error(Errors.ACCESS_DENIED);
-        Response challenge = context.form()
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/authentication/authenticators/access/DenyAccessAuthenticator.java:49]: Isn't possible to keep the current construct/API and be smarter about how the response should be created based on the media type when actually calling `createErrorPage`?

Comment (lexcao) [reply services/src/main/java/org/keycloak/authentication/authenticators/access/DenyAccessAuthenticator.java:49]: Hi @pedroigor 
Sounds a good idea.
Let me try it.

**Comment by pedroigor** [comment]
@lexcao Looks like failures are related. Would you mind checking them?

**Comment by lexcao** [comment]
Hi @pedroigor 
I just fixed the failed test, thanks.
The error was  NPE in `flowPath` called by `getActionUrl` from `context.form()`

```
23:26:53,742 ERROR [org.keycloak.services.error.KeycloakErrorHandler] Uncaught server error
java.lang.IllegalArgumentException: RESTEASY003715: path was null
	at org.jboss.resteasy.specimpl.ResteasyUriBuilderImpl.path(ResteasyUriBuilderImpl.java:355)
	at org.keycloak.authentication.AuthenticationProcessor$Result.getActionUrl(AuthenticationProcessor.java:566)
	at org.keycloak.authentication.AuthenticationProcessor$Result.form(AuthenticationProcessor.java:540)
	at org.keycloak.authentication.authenticators.access.DenyAccessAuthenticator.authenticate(DenyAccessAuthenticator.java:49)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processSingleFlowExecutionModel(DefaultAuthenticationFlow.java:445)
	at org.keycloak.authentication.DefaultAuthenticationFlow.processFlow(DefaultAuthenticationFlow.java:249)
....
```

Since there is no `flowPath` in the token endpoint, is maybe an undefined behavior.
I just hard code (see the PR) to make it with value and seek for your suggestion.

or should we be able to handle the `flowPath` in the `DenyAccessAuthenticator`?
* I do not want to handle `flowPath` in an authenticator

**Comment by pedroigor** [comment]
> or should we be able to handle the `flowPath` in the `DenyAccessAuthenticator`?
> 
> * I do not want to handle `flowPath` in an authenticator

TBH, I need to check. I'll run your PR locally and look at the failing tests to see how it goes.

**Comment by lexcao** [comment]
Thanks.
The failing test is `configureDirectGrantFlowWithDenyAccess` if you remove this line of code
```
// services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java
.setFlowPath("token")
```

File: services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java [hunk 2]
```diff
@@ -678,6 +678,7 @@ public Response resourceOwnerPasswordCredentialsGrant() {
         AuthenticationProcessor processor = new AuthenticationProcessor();
         processor.setAuthenticationSession(authSession)
                 .setFlowId(flowId)
+                .setFlowPath("token")
```
Comment (pedroigor) [inline services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java:681]: Looking at https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java#L544, we should be doing something very similar as you are doing in the `FreemarkerLoginFormsProvider`. If the request is a backend request we don't really need to handle forms.

@mposolda wdyt?

**Review by pedroigor — COMMENTED** [review]
@lexcao Sorry for the long delay. Only a single comment about how we should deal with backend requests when processing forms.

Once we sort it out, LGTM.

**Review by pedroigor — COMMENTED** [review]
@lexcao The failures seem related. Could you check it out?

Talked with @mposolda and the change you did to workaround the flow path issue is OK to move forward.

Thanks.

**Comment by pedroigor** [comment]
@lexcao Btw, you also need to sign the commit by setting the `-s` option when creating te commit.

**Comment by lexcao** [comment]
Hi @pedroigor 
I have updated the PR, maybe we can merge it if the CI is all green.

**Review by pedroigor — COMMENTED** [review]
@lexcao Failure seems related. See https://github.com/keycloak/keycloak/actions/runs/7448885531/job/20264259312?pr=22006#step:3:22425.

**Comment by lexcao** [comment]
@pedroigor 
Thanks, let me try again.

**Comment by lexcao** [comment]
This changes break the tests for SAML POST binding mode like `org.keycloak.testsuite.broker.KcSamlBrokerTest#loginInResponseNoRelayState`.

```
Expected: response body matches a string containing "Unexpected error when authenticating with identity provider"
     but: was <HttpResponseProxy{...} with entity {"error":"identityProviderUnexpectedErrorMessage"}
```

These are requests with POST method and asserting the response body text.
But here we just respond raw error messages with JSON.

I come up with some ways to handle this,
1. Fix the broken test with JSON response assertions, which would break the current response of SAML.
2. Handle the message bundle if the tests assertions failed on the error message, which may be out of scope of this PR.
3. Responding the error only in the `DenyAccessAuthenticator`, which keep the scope.

I am not sure we could break the SAML error response, or the JSON is the correct way?
I prefer the last method.

WDYT? @pedroigor

**Comment by pedroigor** [comment]
> I am not sure we could break the SAML error response, or the JSON is the correct way?

Yeah, we can't break such things here. 

> These are requests with POST method and asserting the response body text.
But here we just respond raw error messages with JSON.

Looks like the main issue here is that we default to returning JSON if no accept header is set. I think we should default to return HTML.

IIRC, the behavior we are adding here is more related to backend requests so for these cases we can enforce the accept header being set if you want a proper content-type as a response.

**Comment by lexcao** [comment]
@pedroigor 
Good idea, let me try it.

**Comment by pedroigor** [comment]
@lexcao Now the test failure is the opposite :) What we expect. We need to update the `OAuthClient` to set the appropriate header when making back-end requests ...

**Comment by pedroigor** [comment]
@lexcao I see ... I think `403` makes more sense too it looks more like an authorization than an authentication check. @mposolda wdyt?

File: services/src/main/java/org/keycloak/forms/login/freemarker/FreeMarkerLoginFormsProvider.java [hunk 3]
```diff
@@ -29,11 +29,8 @@
 import java.util.Map;
 import java.util.Objects;
 import java.util.Properties;
-import jakarta.ws.rs.core.MultivaluedHashMap;
-import jakarta.ws.rs.core.MultivaluedMap;
-import jakarta.ws.rs.core.Response;
-import jakarta.ws.rs.core.UriBuilder;
-import jakarta.ws.rs.core.UriInfo;
+
+import jakarta.ws.rs.core.*;
```
Comment (mabartos) [inline services/src/main/java/org/keycloak/forms/login/freemarker/FreeMarkerLoginFormsProvider.java:33]: @lexcao Please, don't use the star imports.

Comment (lexcao) [reply services/src/main/java/org/keycloak/forms/login/freemarker/FreeMarkerLoginFormsProvider.java:33]: Thanks, let me fix it.

**Comment by lexcao** [comment]
Hi @pedroigor 
I updated the PR with handling the 403 for the `UserSessionLimitsAuthenticator` as well, please help review.

**Review by pedroigor — APPROVED** [review]
@lexcao LGTM, thanks.

File: services/src/main/java/org/keycloak/utils/MediaTypeMatcher.java [hunk 4]
```diff
@@ -1,16 +1,24 @@
...
-        for (jakarta.ws.rs.core.MediaType m : headers.getAcceptableMediaTypes()) {
-            if (!m.isWildcardType() && m.isCompatible(jakarta.ws.rs.core.MediaType.TEXT_HTML_TYPE)) {
+        return isAcceptMediaType(headers, MediaType.TEXT_HTML_TYPE);
+    }
+
+    public static boolean isJsonRequest(HttpHeaders headers) {
+        return isAcceptMediaType(headers, MediaType.APPLICATION_JSON_TYPE);
+    }
+
+    private static boolean isAcceptMediaType(HttpHeaders headers, MediaType textHtmlType) {
```
Comment (mposolda) [inline services/src/main/java/org/keycloak/utils/MediaTypeMatcher.java:16]: minor: renaming the latter variable to `mediaType` might be better (not a blocker for this PR unless there are other points to address).

**Review by mposolda — APPROVED** [review]
@lexcao Nice work! Added one very minor comment inline, but not a blocker (unless there are other comments).

File: services/src/main/java/org/keycloak/utils/MediaTypeMatcher.java (hunk 4 above)
Comment (lexcao) [reply services/src/main/java/org/keycloak/utils/MediaTypeMatcher.java:16]: Thanks for the suggestion. I just forgot this one.
I will take care of the param name next time.
